[pytorch] fewer cuda sync in unique by using cub instead of thrust

jspark1105 opened the pull request:
Summary: Use cub library instead of thrust to reduce # of cuda stream synchronize.

Differential Revision: D28088029

ngimel commented:
@jspark1105 you'll need to add mappings for the cub functions you are using to `cuda_to_hip_mappings.py`

zasdfgbnm commented on aten/src/ATen/native/cuda/Unique.cu:
```diff
@@ -86,6 +88,13 @@ std::tuple<Tensor, Tensor, Tensor> unique_cuda_template(
   auto options = self.options().dtype(kLong);
   Tensor output = self.clone(at::MemoryFormat::Contiguous).reshape(-1);
   int64_t num_inp = output.numel();
+  if (num_inp <= INT_MAX) {
+    // A common case is using CUB that requires fewer cuda stream synchronize
+    // but defined in a separate file because thrust and ATen/cuda/cub.cuh
+    // can't be used at the same compilation unit.
+    return internal::unique_cuda_template_cub<scalar_t>(
+        self, consecutive, return_inverse, return_counts);
+  }
```
How about removing the thrust path completely for the non-dim unique? Because the thrust path does not support `num_inp > INT_MAX` as well.

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+    int64_t n,
+    const scalar_t* input,
+    int* output) {
+  auto begin = blockIdx.x * blockDim.x + threadIdx.x;
+  auto stride = gridDim.x * blockDim.x;
```
Consider using `CUDA_KERNEL_LOOP` in https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/aten/src/ATen/cuda/detail/KernelUtils.h#L21

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+    const int64_t* input,
+    const int64_t* indices,
+    int64_t* output) {
+  auto begin = blockIdx.x * blockDim.x + threadIdx.x;
+  auto stride = gridDim.x * blockDim.x;
+  for (auto i = begin; i < n; i += stride) {
```
Consider using `CUDA_KERNEL_LOOP ` in https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/aten/src/ATen/cuda/detail/KernelUtils.h#L21

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+  if (!return_inverse || num_inp == 0) {
+    inverse_indices = at::empty({0}, options);
+  } else {
+    TORCH_CHECK(
```
```suggestion
    TORCH_INTERNAL_ASSERT(
```

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+        "return_inverse is set to true, but sorted_indices is undefined. Send a bug report!");
+    inverse_indices = at::empty({num_inp}, options);
+    int* inv_loc_ptr =
+        reinterpret_cast<int*>(inverse_indices.data_ptr<int64_t>());
```
Why casting the pointer to `int*`? Is this correct?

zasdfgbnm replied:
If you want to use int, why not creating a `kInt` tensor since the beginning?

jspark1105 replied:
This is a micro-optimization to reuse inverse_indices instead of creating a new one. adjacent_difference_kernel just needs int (actually boolean suffices) but scatter_kernel outputs int64. If you think this is a premature optimization, I'm happy to create a new tensor here :)

zasdfgbnm replied:
I see, this is a good optimization. You can keep it.

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+  Tensor counts = at::empty({0}, options);
+  Tensor length = at::empty({1}, options);
+  int64_t num_out;
+  size_t temp_storage_bytes = 0;
```
Is this used?

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+        <<<grid, block, 0, stream>>>(num_inp, data, inv_loc_ptr);
+    C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+    size_t temp_storage_bytes = 0;
```
Is this used?

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+        length.data_ptr<int64_t>(),
+        num_inp,
+        stream);
+    num_out = *length.cpu().data_ptr<int64_t>();
```
```suggestion
    num_out = length.item<int64_t>();
```

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+        length.data_ptr<int64_t>(),
+        num_inp,
+        stream);
+    num_out = *length.cpu().data_ptr<int64_t>();
```
```suggestion
    num_out = length.item<int64_t>();
```

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+  return std::tuple<Tensor, Tensor, Tensor>(output, inverse_indices, counts);
+}
+
+#define INSTANTIATE_UNIQUE_CUDA_TEMPLATE(TYPE)                                                     \
```
Remove spaces before `/` to make it aligned.

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+
+    size_t temp_storage_bytes = 0;
+    Tensor inv_loc_out = at::empty({num_inp}, options);
+    CUB_WRAPPER(
+        cub::DeviceScan::InclusiveSum,
+        inv_loc_ptr,
+        inv_loc_out.data_ptr<int64_t>(),
+        num_inp,
+        stream);
+
+    scatter_kernel<<<grid, block, 0, stream>>>(
+        num_inp,
+        inv_loc_out.data_ptr<int64_t>(),
+        sorted_indices.data_ptr<int64_t>(),
+        inverse_indices.data_ptr<int64_t>());
```
Looks like for the case of `consecutive  == true`, `sorted_indices` will be `0, 1, 2, 3, ...`, therefore the `scatter_kernel` is effectively a memcpy? You can add an if statement for this case and make `InclusiveSum` directly output to `inverse_indices.data_ptr<int64_t>()`.

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+      cuda::cub::sort_keys(self.data_ptr<scalar_t>(), sorted_data, num_inp);
+    }
+  } else {
+    sorted_indices = at::arange(0, num_inp, options);
```
We no longer need this when `consecutive` is true, so I believe you can move the `arange` call inside the if clause below

zasdfgbnm commented on aten/src/ATen/native/cuda/UniqueCub.cu:
```diff
@@ -0,0 +1,201 @@
+    }
+  } else {
+    if (!consecutive) {
+      sorted_indices = at::arange(0, num_inp, options);
+      Tensor sorted_indices_temp = at::empty({num_inp}, options);
+      cuda::cub::sort_pairs(
+          self.data_ptr<scalar_t>(),
+          sorted_data,
+          sorted_indices.data_ptr<int64_t>(),
+          sorted_indices_temp.data_ptr<int64_t>(),
+          num_inp);
+      sorted_indices = sorted_indices_temp;
```
This code can be simplified as
```suggestion
      Tensor range = at::arange(0, num_inp, options);
      sorted_indices = at::empty({num_inp}, options);
      cuda::cub::sort_pairs(
          self.data_ptr<scalar_t>(),
          sorted_data,
          range.data_ptr<int64_t>(),
          sorted_indices.data_ptr<int64_t>(),
          num_inp);
```

zasdfgbnm reviewed (approved):
Very great job! Thank you for working on this. I have one final tiny improvement suggestion.

jspark1105 committed:
[pytorch] fewer cuda sync in unique by using cub instead of thrust (#57323)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/57323

Use cub library instead of thrust to reduce # of cuda stream synchronize.

Reviewed By: ngimel

Differential Revision: D28088029

fbshipit-source-id: 6e51ab8771ae6f7f993136dac6cb12628e4d1514
